Split apart runc shim into plugin components

dmcgowan opened the pull request:
Still some remaining items to work on here such as
- Moving task service component
- Registering plugins

thaJeztah commented on runtime/v2/runc/v2/service.go:
```diff
@@ -21,61 +21,71 @@ package v2
-}
+// New returns a new shim service that can be used via GRPC
+// TODO(2.0): Remove this function, rely on plugin registration
+func New(_ context.Context, id string, _ shim.Publisher, _ func()) (shim.Shim, error) {
```
Just glancing over the changes, and I'm a bit confused; is this backward compatible (as in; all arguments are now ignored, whereas previously they did something).

If the change is non-backward compatible from that perspective (even without the signature changing), I'm wondering if instead we should explicitly break it, and remove the arguments?

It doesn't look to be widely used (https://grep.app/search?q=github.com/containerd/containerd/runtime/v2/runc/v2), so mostly commenting (wondering if we want to stick with SemVer for these), but I see at least firecracker is using this; (cc @kzys) https://github.com/firecracker-microvm/firecracker-containerd/blob/568c84051caf61c1fc77a0b62171c22c40219b6a/agent/service.go#L99

dmcgowan replied:
Not compatible for that firecracker use case. I opened this PR earlier so I can test that the shutdown logic is working. I can probably leave that function compatible though and just avoid it entirely for our own compilation. When we use it in the future, it can just register the task service. The update strips the task service from the "Shim", just leaving the shim manager components. The linked to use relies on the task service though. There is no real way to keep the interfaces clean and separate shim manager and task service though. The "2.0" comment will give a chance to completely clean that up though.

thaJeztah replied:
> Not compatible for that firecracker use case. I opened this PR earlier so I can test that the shutdown logic is working. I can probably leave that function compatible though and just avoid it entirely for our own compilation. 

Thanks for clarifying; yes if we could keep it compatible (at least for now), that would be great

dmcgowan commented:
Updated to restore backwards compatibility with existing `Init` func implementations. Added a new runner function which does not require the init function at all in favor of plugins and manager implementation. Still using the `Init` function for the runc plugin in latest change to see tests passing before switching over.

dmcgowan commented:
Marking this as ready to review, I recommend starting with `runtime/v2/shim/shim.go`

crosbymichael reviewed (approved):
LGTM

AkihiroSuda commented on runtime/v2/runc/manager/manager.go:
```diff
@@ -0,0 +1,279 @@
+// +build linux
```
Not needed ( git mv manager.go manager_linux.go )

fuweid reviewed (approved):
LGTM
